File: coinChange.R
# R script that takes an integer in between 1 to 99 as the input 
# It returns the number of quarters, dimes, nickles, and pennies 
# The sum of the values of coins is equal to the input integer
#
#
# To run the script, type in the shell:
# Rscript choinChange.R  99
#
#
# Auxiliary Function to do the calculation  
processChange <- function(balance,denom){
			numOfCoin <- balance%/%denom      # R floor division 
	      		remainBal <- balance%%denom       # R modulous
			return (c(numOfCoin, remainBal))
			}

# Main Function 
makeChange <- function(amountChange){
	   # Create a list 
	   coinName   <- c('quarter','dime','nickle','penny') 
    	   coinVal    <- c(25,10,5,1)
	   changeList <- list(key=coinName, val=coinVal)

    	   # Create empty vectors to save the results 
    	   a <- vector()
    	   b <- vector()

	   # Loop over the changeList
    	   # processChange returns a two elements vector 
    	   # the second element of the vector is the residual
    	   # assign the residual to the variable amountChange
    	   # append the number of coins to the empty list 

	   d = length(changeList$key)
    	   for (i in 1:d){
               res = processChange(amountChange, changeList$val[i])  
               amountChange = res[2]     
              
               a[i] <- res[1] 
	       b[i] <- coinName[i] #changeList$key[i])         
 	    }

	    coinList <- list(Num=a, Name=b)
	    return (coinList)
}



argv <- commandArgs(TRUE)
amountToChange <- as.numeric(argv[1])
cat('\n')
cat('The amount to change is,', amountToChange, ', cents.','\n')
cat('The output is:', '\n')
coinList = makeChange(amountToChange)

cat('[')
for (i in 1:length(coinList$Num)){
    cat('(', coinList$Num[i], ', ', coinList$Name[i], '),', sep='')
}
cat(']', '\n')

File: coinChange.py
# Python script that takes an integer in between 1 to 99 as the input 
# It returns the number of quarters, dimes, nickels, and pennies 
# The sum of the values of coins is equal to the input integer 
#
#
# import sys package for command line input 
# import collections package for Ordered Dictionary
import sys, collections 


# Auxiliary Function to do the calculation  
def processChange(balance,denom):
    numOfCoin = balance/denom           # python integer division 
    remainBal = balance%denom           # python modulus 
    return (numOfCoin, remainBal)


# Main Function 
def makeChange(amountChange):
    # Create an ordered list 
    changeList= collections.OrderedDict()
    changeList['quarter']=25
    changeList['dime']=10
    changeList['nickel']=5
    changeList['penny']=1

    # Create an empty list to save the results 
    coinList=[]
    
    # Loop over the ordered list
    # processChange returns a two elements tuple
    # the second element of the tuple is the residual
    # assign the residual to the variable amountChange
    # append the number of coins to the empty list 
    for key,val in changeList.items():
        res = processChange(amountChange, val)  
        amountChange = res[1]                   
        coinList.append((res[0], key))          
    return coinList


if __name__=='__main__':
    amountToChange = int(sys.argv[1])
    print ''
    print 'The amount to change is,', amountToChange, ', cents.'
    print 'The output is:'
    coinList = makeChange(amountToChange)
    print coinList
